### ISSUE TITLE
Processing MLF files

### ISSUE THREAD

**Issue by jjatria** [badner/lab_to_praat#11]
I was looking into the processing of Master Label Files (are these used by HTS as well?), thinking that they were just a convenient way of grouping regular label files. I thought, if that's the case, then reading an MLF should just generate `TextGrid` objects for all the contained labels.

But it turns out that, rather than a sort of collection of labels, it is a directory of label files: it maps specific label file name patterns to paths where files with matching names should be findable.

In that sense, it may not be desirable to generate all the contained annotations, because these may be way too many.

So, I can think of two possible approaches to implementing MLF file processing:

1. We follow the original plan, and treat the MLF file as a collection of annotations. If given an MLF like

    ~~~
    #!MLF!#
    "*" -> "/db/dr1/labs"
    "*" -> "/db/dr2/labs"
    ~~~

    then this would result in the generation of as many `TextGrid` files as there are labels matching those patterns (`"*"`) in those directories (or as many alternatives there are in those label files).

2. We treat MLF files as _something you can run queries against_, which is what I understand them to be really. In this case, reading the above file would not result in any `TextGrid` at all, but in some sort of representation of that mapping (which I think could be a `Table`). With that mapping selected, then you could run a query. 

    Reading about the [MLF search algorithm](http://www1.icsi.berkeley.edu/Speech/docs/HTKBook3.2/node100_mn.html), it seems that never more than one label is matched, because search stops after the first match. So running something like

    ~~~
    runScript: "read_mlf.praat", "path/to/file.mlf"
    mlf = selected("Table")
    runScript: "query_mlf.praat", "test*"
    # Or perhaps a call to a procedure, either expecting a selected Table 
    @mlf_query: "test*"
    # or requiring an ID to work on
    @mlf_query: mlf, "test*"
    ~~~

    would then result in the search for a match to the `"test*"` pattern in the directories specified in the MLF file, and the generation of the `TextGrid` objects resulting from the first matching label file or direct label. If none is found, an error is raised.

    This all seems doable. As for the internal representation of MLF files, I'm thinking a `Table` object with the following columns should be enough:

    * `pattern`, with the pattern to match
    * `mode`, with a string for either `full`, `simple`, or `direct` to specify how the data is provided
    * `target`, which can be a path, for `full` and `simple` modes, or the text of the actual label file, for the `direct` mode.

    Indeed, once this is done, this sort of mapping can be used for any kind of files, since it is simply a mapper between name patterns and subdirectories. Once a match is found, the MLF query script could check if it is a label file (extension based?) and open it with the label reader, or pass it to the `Read from file...` Praat command, so that Praat objects can be read as well. This, I think, would be _very_ useful to have.

**Comment by jjatria** [comment]
As a possible improvement, we could implement optional caching, so that if a file has been queried, found, and opened once, it is not necessary to do the process again. 

This would be easily implemented by adding a new row, at the top of the Table, with the exact file name as pattern, the exact path to the file as `target`, and `simple` as the mode. This would still have the file I/O, but it would save subsequent possibly expensive file searches whose result we already know.

**Comment by CShulby** [comment]
no the .mlf has timestamps.  there is an .scp which ahs the paths.  Where
did you find this info because something is confused.  I will annex a
sample .mlf and .scp file I have here.

The .mlf file is very similar to the .lab file.  The parser needs to be
updated.  The biggest problem is as you pointed out, the TextGrid can hold
more information which could be lost in the transfer to .mlf.  If you stick
with just the phonetic level then we have no problems but once praat starts
adding tiers it becomes difficult.

2017-02-02 10:54 GMT-02:00 José Joaquín Atria <notifications@github.com>:

**Comment by jjatria** [comment]
What I wrote above comes from the HTK book, and more precisely from the URL I linked to. There is always the possibility, like you said, that the information is out of date. But from what it says there, at least, I get that the MLF _can_ have timestamps (what I refer to as the `direct` mode), but it _can_ also have simply paths to subdirectories. It says so right there in the [syntax definition](http://www1.icsi.berkeley.edu/Speech/docs/HTKBook3.2/node99_mn.html):

    MLF  = 
        "#!MLF!#"
        MLFDef { MLFDef }
    
    MLFDef =
        ImmediateTranscription | SubDirDef

    ImmediateTranscription = 
        Pattern 
        Transcription 
        "."

    SubDirDef =
        Pattern SearchMode String 

    SearchMode =
        "->" | "=>"

    Pattern =
        String

So, regardless of whether the MLF has an `ImmediateTranscription` or a `SubDirDef`, in both cases there is a `Pattern` that has to match. From what it says there, the MLF file is basically a mapping from file name patterns to transcriptions, which can either be findable in one of the specified subdirectories, or provided directly. What about this is wrong?

I do not know about the SCP files, but from what it says in the HTK book these are "script" files. These sound to me as quite different. Are they not?
